examples/algorithm_progression.py

```diff
@@ -0,0 +1,182 @@
+        'sources': [],
+        'uniformSources': True,
+    }
+    param_desc = [
+        f'{VARIABLE_LAYERS[i]} represents change in {n} as an index of {p}'
+        if len(p) > 1 and i < len(VARIABLE_LAYERS)
+        else f'{n} remains unchanged per run with a value of {p[0]}'
+        for i, (n, p) in enumerate(input_params.items())
+    ]
+    if len(param_desc) > 0:
+        yaml_dict['description'] += f', where {",".join(param_desc)}'
+
+    param_space_combos = list(itertools.product(*input_params.values()))
+    print(f'Beginning {len(param_space_combos)} runs on {max_workers} workers...')
+    num_done = 0
+    with concurrent.futures.ProcessPoolExecutor(max_workers=max_workers) as executor:
+        # cannot pass source through submitted task; it is unpickleable
+        futures = [
+            executor.submit(
+                apply_algorithm,
+                algorithm,
+                input_filename,
+                output_dir,
+                combo,
+                input_params
+            )
+            for combo in param_space_combos
+        ]
+        for future in concurrent.futures.as_completed(futures):
+            num_done += 1
+            print(f'Completed {num_done} of {len(param_space_combos)} runs.')
+            yaml_dict['sources'].append(future.result())
+
+    with open(Path(output_dir, 'results.yml'), 'w') as f:
+        yaml.dump(
+            yaml_dict,
+            f,
+            default_flow_style=False,
+            sort_keys=False,
+        )
+
+
+if __name__ == '__main__':
+    argparser = create_argparser()
+    args = argparser.parse_args()
+
+    algorithm_code = args.algorithm_code
+    input_filename = args.input_filename
+    output_dir = args.output_dir
+    max_workers = args.num_workers
+    if args.param:
+        input_params = {
+            p[0].split(',')[0]: [i.strip() for i in p[0].split(',')[1:]]
+            for p in args.param
+        }
+    else:
+        input_params = {}
+    input_params = {
+        key: np.linspace(
+            float(value[0]), float(value[1]), int(value[2]), endpoint=len(value) > 3
+        )
+        for key, value in input_params.items()
+        if len(value) >= 3
+    }
+
+    if not Path(input_filename).exists():
+        raise ValueError(f'Cannot locate file {input_filename}.')
+
+    algorithm = algorithms.ALGORITHM_CODES[algorithm_code]
+    sig = inspect.signature(algorithm)
+
+    if not output_dir:
+        output_dir = f'{algorithm.__name__}_output'
+        i = 0
+        while Path(output_dir).exists():
+            i += 1
+            output_dir = f'{algorithm.__name__}_output_{i}'
+        os.mkdir(output_dir)
+    else:
+        if not Path(output_dir).exists():
+            os.mkdir(output_dir)
+
+    params = {
+        param.name: (
+            input_params[param.name]
+            if param.name in input_params
+            else algorithms.ALGORITHM_DEFAULT_PARAM_SPREADS[algorithm_code][param.name]
```

Comment:
When we want to sweep just one parameter (say hue_value in the PPC algorithm), we currently have to explicitly say to use a single value for other default params.  I wonder if it would be better to skip using the defaults if any input parameters are given.  @annehaley What do you think?  Otherwise, this is working as desired.

Reply:
Ok I can do that! hanged in [98a33e1](https://github.com/girder/large_image/pull/1051/commits/98a33e194ca9b8d1d95b13880bba24202bb122cd)